xwiki-platform-core/xwiki-platform-icon/xwiki-platform-icon-rest/xwiki-platform-icon-rest-default/src/main/java/org/xwiki/icon/internal/DefaultIconThemesResource.java

```diff
@@ -0,0 +1,146 @@
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.icon.internal;
+
+import java.util.List;
+import java.util.Map;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.Response;
+
+import org.apache.commons.lang3.StringUtils;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.icon.IconException;
+import org.xwiki.icon.IconManager;
+import org.xwiki.icon.IconSet;
+import org.xwiki.icon.IconSetManager;
+import org.xwiki.icon.rest.IconThemesResource;
+import org.xwiki.icon.rest.model.jaxb.Icon;
+import org.xwiki.icon.rest.model.jaxb.Icons;
+import org.xwiki.icon.rest.model.jaxb.ObjectFactory;
+import org.xwiki.model.ModelContext;
+import org.xwiki.model.reference.EntityReference;
+import org.xwiki.model.reference.WikiReference;
+import org.xwiki.rest.XWikiRestComponent;
+
+import static org.xwiki.icon.IconManager.META_DATA_CSS_CLASS;
+import static org.xwiki.icon.IconManager.META_DATA_ICON_SET_TYPE;
+import static org.xwiki.icon.IconManager.META_DATA_URL;
+
+/**
+ * Default implementation of {@link IconThemesResource}.
+ *
+ * @version $Id$
+ * @since 13.4RC1
+ */
+@Component
+@Named("org.xwiki.icon.internal.DefaultIconThemesResource")
+@Singleton
+public class DefaultIconThemesResource implements IconThemesResource, XWikiRestComponent
+{
+    @Inject
+    private ModelContext modelContext;
+
+    @Inject
+    private IconManager iconManager;
+
+    @Inject
+    private IconSetManager iconSetManager;
+
+    @Override
+    public Icons getIconsByTheme(String wikiId, String iconTheme, List<String> names)
+    {
+        return iconsByTheme(wikiId, iconTheme, names);
+    }
+
+    @Override
+    public Icons getIcons(String wikiName, List<String> names)
+    {
+        return iconsByTheme(wikiName, null, names);
+    }
+
+    private Icons iconsByTheme(String wikiId, String iconTheme, List<String> iconNames)
+    {
+        // Save the current entity reference. It will be restored on the method is finished.
+        EntityReference oldEntityReference = this.modelContext.getCurrentEntityReference();
+        try {
+            // Set the requested wiki.
+            this.modelContext.setCurrentEntityReference(new WikiReference(wikiId));
+
+            IconSet iconSet = getIconSet(iconTheme);
+            if (iconSet == null) {
+                throw new WebApplicationException(Response.Status.NOT_FOUND);
+            }
+            String iconSetName = iconSet.getName();
+            ObjectFactory objectFactory = new ObjectFactory();
+            Icons icons = objectFactory.createIcons();
+            List<String> knownIconNames = this.iconManager.getIconNames();
```

Comment:
Small optimization: use a Set in order to speed up the ``contains`` call below.

Reply:
That's a good point. WDYT of providing `getIconNameSet` methods in `DefaultIconManager` and `IconSet`, since the initial datastructure holding the names is a `HashMap` (https://github.com/xwiki/xwiki-platform/blob/08eadbf371e31d3af0aeeb2e0b0b0f849fc447bf/xwiki-platform-core/xwiki-platform-icon/xwiki-platform-icon-api/src/main/java/org/xwiki/icon/IconSet.java#L257). It would be a shame to seet a `Set` converted to a `ArrayList` to be converted again in an `HashSet`. The cost of the conversions might outweight the gain of the contains in some cases!
Another solution could be to provide a `boolean iconNameExists(String iconName)` in `DefaultIconManager` and `IconSet`.

WDYT?

Reply:
+1 for adding ``hasIcon(String iconName)`` (shorter :) )

Reply:
Done